kimchi/src/tests/keccak.rs

```diff
@@ -1,54 +1,237 @@
+    fn to_u64<F: Field>(elem: F) -> u64 {
+        let mut bytes = FieldHelpers::<F>::to_bytes(&elem);
+        bytes.reverse();
+        bytes.iter().fold(0, |acc: u64, x| (acc << 8) + *x as u64)
+    }
+    fn print_line(state: &[u64]) {
+        print!("         ");
+        for x in 0..5 {
+            let quarters = &state[4 * x..4 * (x + 1)];
+            let word = compose(&collapse(&reset(&shift(quarters))));
+            print!("{:016x} ", word);
+        }
+        println!();
+    }
+    fn print_matrix(state: &[u64]) {
+        for x in 0..5 {
+            print!("         ");
+            for y in 0..5 {
+                let quarters = &state[4 * (5 * y + x)..4 * (5 * y + x) + 4];
+                let word = compose(&collapse(&reset(&shift(quarters))));
+                print!("{:016x} ", word);
             }
-            assert_eq!(
-                PallasField::from(state[x][y].rotate_left(bits)),
-                witness[1][1 + 2 * rot],
-            );
-            rot += 1;
+            println!();
         }
     }
+
+    let row = witness
+        .iter()
+        .map(|x| to_u64::<F>(x[round]))
+        .collect::<Vec<u64>>();
+    let next = witness
+        .iter()
+        .map(|x| to_u64::<F>(x[round + 1]))
+        .collect::<Vec<u64>>();
+
+    println!("----------------------------------------");
+    println!("ROUND {}", round);
+    println!("State A:");
+    print_matrix(&row[0..100]);
+    println!("State C:");
+    print_line(&row[100..120]);
+    println!("State D:");
+    print_line(&row[320..340]);
+    println!("State E:");
+    print_matrix(&row[340..440]);
+    println!("State B:");
+    print_matrix(&row[1440..1540]);
+
+    let mut state_f = row[2340..2344].to_vec();
+    let mut tail = next[4..100].to_vec();
+    state_f.append(&mut tail);
+
+    println!("State F:");
+    print_matrix(&state_f);
+    println!("State G:");
+    print_matrix(&next[0..100]);
+}
+
+// Sets up test for a given message and desired input bytelength
+fn test_keccak<G: KimchiCurve>(message: BigUint) -> BigUint
+where
+    G::BaseField: PrimeField,
+{
+    let bytelength = message.to_bytes_be().len();
+    let padded_len = {
+        let mut sized = message.to_bytes_be();
+        sized.resize(bytelength - sized.len(), 0);
+        pad(&sized).len()
+    };
+    let _index = create_test_constraint_system::<G>(padded_len);
+    let witness = create_keccak_witness::<G>(message);
+
+    for r in 1..=24 {
+        print_witness::<G::ScalarField>(&witness, r);
+    }
+
+    let mut hash = vec![];
+    let hash_row = witness[0].len() - 2; // Hash row is dummy row
+    println!();
+    println!("----------------------------------------");
+    print!("Hash: ");
+    for b in 0..32 {
+        hash.push(FieldHelpers::to_bytes(&witness[200 + b][hash_row])[0]);
+        print!("{:02x}", hash[b]);
+    }
+    println!();
+    println!();
+
+    BigUint::from_bytes_be(&hash)
+}
+
+#[test]
```

Comment:
I think there are also some property based testing we could apply.
We could generate random inputs and verify the properties are respected.